doc/reference/reference_lua/box_null.rst

```diff
@@ -0,0 +1,138 @@
+.. _box-null:
+
+-------------------------------------------------------------------------------
+                             Constant `box.NULL`
+-------------------------------------------------------------------------------
+
+There are some major problems with using a Lua **nil** value in tuples.
+For example: you can't correctly assess length of sparse array (an array with
+nullable values set to NULL) in case of using Lua **nil** value.
+
+**Example:**
+
+.. code-block:: tarantoolsession
+
+    tarantool> t = {0, nil, 1, 2, nil}
+    ---
+    ...
+
+    tarantool> t
+    ---
+    - - 0
+      - null
+      - 1
+      - 2
+    ...
+
+    tarantool> #t
+    ---
+    - 4
+    ...
+
+Provided example clearly shows a problem with sparse arrays.
+
+.. NOTE::
+
+    Trying to find length for sparse arrays in LuaJIT leads to
+    `undefined behaviour <https://www.lua.org/manual/5.2/manual.html#3.4.6>`_.
+
+To overcome this and some other issues use ``box.NULL`` constant instead of **nil**.
+
+Using box.NULL
+--------------
+
+``box.NULL`` is a value of cdata type representing a NULL pointer.
+It is similar to ``msgpack.NULL``, ``json.NULL`` and ``ffi.NULL``. So it is
+some not **nil** value, even if it is a pointer to NULL.
+
+Use ``box.NULL`` only with capitalized NULL (``box.null`` is incorrect).
+
+.. NOTE::
+
+    Technically speaking ``box.NULL`` equals to ``ffi.cast('void *', 0)``.
+
+**Example:**
+
+.. code-block:: tarantoolsession
+
+    tarantool> t = {0, box.NULL, 1, 2, box.NULL}
+    ---
+    ...
+
+    tarantool> t
+    ---
+    - - 0
+      - null
+      - 1
+      - 2
+      - null
+    ...
+
+    tarantool> #t
+    ---
+    - 5
+    ...
+
+.. NOTE::
+
+    There is a possible problem induced by using ``box.NULL``.
+    Avoid using implicit comparisons with nullable values when using ``box.NULL``.
+    Due to `Lua behaviour <https://www.lua.org/manual/5.1/manual.html#2.4.4>`_
+    returning from condition expression anything except **false** or **nil**
+    is considered as **true**. And as it was mentioned earlier ``box.NULL`` is a
+    pointer by design.
+
+    That is why expression ``box.NULL`` will always be considered true in case
+    it is used as condition in comparison. This means that code
+
+    ``if box.NULL then func() end``
+
+    will always execute function ``func()`` (because condition ``box.NULL`` will
+    always be neither **false** nor **nil**).
+
+Distinction of nil and box.NULL
+-------------------------------
+
+If condition expression ``x == nil`` is **true** the ``x`` could be **nil** or
+``box.NULL``.
```

Comment:
```suggestion
Use expression ``if x == nil`` to check if the ``x`` is either **nil** or ``box.NULL``.
```

Reply:
This is basically the same, but instead of description of whats going on you give suggestion on how to behave. No clear description of result interpretation given that way.

Reply:
check with another reviewer